src/components/tl/cuda/bcast/bcast_linear.c

```diff
@@ -0,0 +1,423 @@
+    case STAGE_INIT_BAR_ROOT:
+        st = root_find_free_barrier(task);
+        if (st == UCC_OK) {
+            task->bcast_linear.stage = STAGE_SYNC;
+        } else if (st != UCC_ERR_NOT_FOUND) {
+            task->super.status = st;
+        }
+        // no free barriers found, try next time
+        return;
+    case STAGE_FIND_BAR_PEER:
+        st = peer_find_free_barrier(task);
+        if (st == UCC_OK) {
+            // barrier found, continue to next stages
+            task->bcast_linear.stage = STAGE_SYNC;
+        } else if (st != UCC_ERR_NOT_FOUND) {
+            task->super.status = st;
+        }
+        // no free barriers found by root, try next time
+        return;
+    case STAGE_SYNC:
+        if (ucc_tl_cuda_get_sync_root(task, task->bcast_linear.root) != UCC_OK) {
+            return;
+        }
+        task->bcast_linear.step = 0;
+        st                      = ucc_tl_cuda_bcast_linear_setup_start(task);
+        if (st != UCC_OK) {
+            task->super.status = st;
+            return;
+        }
+        task->bcast_linear.stage = STAGE_SETUP;
+    case STAGE_SETUP:
+        st = ucc_tl_cuda_bcast_linear_setup_test(task);
+        if (st != UCC_OK) {
+            task->super.status = st;
+            return;
+        }
+        if (trank == task->bcast_linear.root) {
+            task->bcast_linear.stage = STAGE_COPY;
+        } else {
+            task->bcast_linear.stage = STAGE_WAIT_ROOT;
+        }
+    default:
+        break;
+    }
+
+    if (trank == task->bcast_linear.root) {
+        // Root scenario
+        // fall-through between cases is intentional
+        switch (task->bcast_linear.stage) {
+        case STAGE_COPY:
+            // copy from src buffer to scratch
+            dbuf = PTR_OFFSET(TASK_SCRATCH(task, trank),
+                              task->bcast_linear.step % 2 * half_scratch_size);
+            sbuf = PTR_OFFSET(task->bcast_linear.sbuf, offset_buff);
+            st   = ecopy(dbuf, sbuf, chunk_size, exec,
+                         &task->bcast_linear.exec_task);
+            if (st != UCC_OK) {
+                ucc_error("failed to post ecopy task");
+                task->super.status = st;
+                return;
+            }
+            task->bcast_linear.stage = STAGE_WAIT_COPY;
+        case STAGE_WAIT_COPY:
+            etask = task->bcast_linear.exec_task;
+            ucc_assert(NULL != etask);
+            st = ucc_ee_executor_task_test(etask);
+            if (st == UCC_OK) {
+                ucc_ee_executor_task_finalize(etask);
+                task->bcast_linear.exec_task = NULL;
+                // signal others
+                ++task->bcast_linear.step;
+                set_rank_step(task, task->bcast_linear.root,
+                              task->bcast_linear.step, 0);
+                task->bcast_linear.stage = STAGE_WAIT_ALL;
+            } else {
+                // not ready
+                return;
+            }
```

Comment:
Better readability to treat the `st != UCC_OK` case before (using that we can return), and also we need to handle if the executor is in error state
```suggestion
            if (st != UCC_OK) {
                task->super.status = st;
                return;
            }
            ucc_ee_executor_task_finalize(etask);
            task->bcast_linear.exec_task = NULL;
            // signal others
            ++task->bcast_linear.step;
            set_rank_step(task, task->bcast_linear.root,
                          task->bcast_linear.step, 0);
            task->bcast_linear.stage = STAGE_WAIT_ALL;
```